### PR TITLE
Introduce new Material 3 `Slider` shapes

### LINKED ISSUE #141842: Update `Slider` for Material 3 redesign
M3 sliders have an [updated look](https://m3.material.io/components/sliders/overview)

image
image

Since this occurred after the primary Material 3 transition, this could potentially be an option that starts as defaulting to `false`, switches to `true`, and then is eventually removed. Or a new widget as they are significantly different.

**Specification:** https://m3.material.io/components/sliders/specs
**December 2023 updates:** https://m3.material.io/components/sliders/overview#79de9466-a59c-4236-9099-735e67a38e53

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [Update `Slider` for Material 3 redesign](https://github.com/flutter/flutter/issues/141842)
previous implementation https://github.com/flutter/flutter/pull/147783

### Description

This PR introduces new Material 3 Slider design. 

### Slider Preview

Screenshot 2024-07-24 at 16 38 11

Screenshot 2024-07-24 at 16 38 24

### Value indicator Preview

https://github.com/user-attachments/assets/45fa001c-de81-433a-a8e9-6c0d6a2335c0

### New stop indicator

https://github.com/user-attachments/assets/ad05621d-042d-4b17-9dbb-7f7b802a2593

### Customized 

Screenshot 2024-07-24 at 16 41 49

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by QuncCccccc — COMMENTED** [review]
Thanks for adding the feature! This looks so great! I haven't finish reviewing the whole change, will continue tomorrow😎!

I also noticed slider.0.dart and slider.1.dart are basically the same. I think slider.1.dart was added when `useMaterial3` was still false? Maybe we can delete one of these and add a button to switch between M2 and M3 in just one example app. This can be done later in a separate PR:)

File: packages/flutter/lib/src/material/slider.dart [hunk 1]
```diff
@@ -2018,25 +2072,26 @@ class _SliderDefaultsM3 extends SliderThemeData {
   Color? get disabledInactiveTrackColor => _colors.onSurface.withOpacity(0.12);
 
   @override
-  Color? get disabledSecondaryActiveTrackColor => _colors.onSurface.withOpacity(0.12);
+  Color? get disabledSecondaryActiveTrackColor => _colors.onSurface.withOpacity(0.38).withOpacity(0.12);
```
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:2075]: Yes, good catch!

File: packages/flutter/lib/src/material/slider.dart [hunk 2]
```diff
@@ -2018,25 +2072,26 @@ class _SliderDefaultsM3 extends SliderThemeData {
...
 
   @override
   Color? get thumbColor => _colors.primary;
 
   @override
-  Color? get disabledThumbColor => Color.alphaBlend(_colors.onSurface.withOpacity(0.38), _colors.surface);
+  Color? get disabledThumbColor => _colors.onSurface.withOpacity(0.38).withOpacity(0.38);
 
   @override
   Color? get overlayColor => MaterialStateColor.resolveWith((Set<MaterialState> states) {
```
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:2097]: You're right, I also tested Slider on the native and it doesn't seem to have overlay color anymore but when using it there is no other way to indicate if the Slider can is hovered or focused. I sent a Demo on Discord.

IMO, to avoid breaking existing tests and focus, hover behavior of the Flutter Slider widget, we can keep the overlay color as it is until we have other ways to indicate these states.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 3]
```diff
@@ -588,6 +590,20 @@ class SliderThemeData with Diagnosticable {
   /// If specified, overrides the default value of [Slider.allowedInteraction].
   final SliderInteraction? allowedInteraction;
 
+  /// The size of the [BarSliderThumbShape] thumb.
+  ///
+  /// If [SliderThemeData.thumbShape] is [BarSliderThumbShape], this property is used to
+  /// set the size of the thumb. Otherwise, the default thumb size is 4 pixels for the
+  /// width and 44 pixels for the height.
+  final MaterialStateProperty<Size?>? barThumbSize;
```
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:598]: The reason i made it `barThumbSize` because it's a `Size` type and this is not compatible with existing Thumb as it is just a circle and it would be `double` aka radius value. If the user expect `thumbSize` to update existing thumb it might create confusion. 

However, you rasied a good point on consistency. I will take a look if we `Size` `thumbSize` can be used for existing thumb radius by rounding to square value.

File: packages/flutter/lib/src/material/slider.dart [hunk 4]
```diff
@@ -1932,12 +1943,14 @@ class _RenderValueIndicator extends RenderBox with RelayoutWhenSystemFontsChange
 }
 
 class _SliderDefaultsM2 extends SliderThemeData {
-  _SliderDefaultsM2(this.context)
-    : _colors = Theme.of(context).colorScheme,
+  _SliderDefaultsM2({ required this.context, required SliderThemeData sliderTheme })
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider.dart:1946]: Can the SliderThemeData not be acquired like the ThemeData is below using the context?

File: packages/flutter/lib/src/material/slider.dart [hunk 5]
```diff
@@ -1985,6 +1998,47 @@ class _SliderDefaultsM2 extends SliderThemeData {
...
+  @override
+  Color? get valueIndicatorColor {
+    if (_sliderTheme.valueIndicatorShape is RoundedRectSliderValueIndicatorShape) {
+      return _colors.inverseSurface;
+    }
+    return _colors.primary;
+  }
+
+  @override
+  MaterialStateProperty<Size?>? get barThumbSize =>
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider.dart:2023]: Nit: Remove `=>` when followed by a multiline block.

File: packages/flutter/lib/src/material/slider.dart [hunk 6]
```diff
@@ -1997,7 +2051,7 @@ class _SliderDefaultsM2 extends SliderThemeData {
 
 class _SliderDefaultsM3 extends SliderThemeData {
   _SliderDefaultsM3(this.context)
-    : super(trackHeight: 4.0);
+    : super(trackHeight: 16.0);
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider.dart:2054]: Why the change here, do we expect it to be breaking? Is this value represented by a token?

File: packages/flutter/lib/src/material/slider.dart [hunk 7]
```diff
@@ -2018,25 +2072,26 @@ class _SliderDefaultsM3 extends SliderThemeData {
...
-  Color? get disabledSecondaryActiveTrackColor => _colors.onSurface.withOpacity(0.12);
+  Color? get disabledSecondaryActiveTrackColor => _colors.onSurface.withOpacity(0.38);
 
   @override
-  Color? get activeTickMarkColor => _colors.onPrimary.withOpacity(0.38);
+  Color? get activeTickMarkColor => _colors.secondaryContainer;
 
   @override
-  Color? get inactiveTickMarkColor => _colors.onSurfaceVariant.withOpacity(0.38);
+  // TODO(tahatesser): Update this hard-coded value to use the correct token value.
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider.dart:2081]: Can you file an issue to follow up and include it here?

File: packages/flutter/lib/src/material/slider.dart [hunk 8]
```diff
@@ -2054,12 +2109,48 @@ class _SliderDefaultsM3 extends SliderThemeData {
...
+  SliderComponentShape? get thumbShape => const BarSliderThumbShape();
+
+  @override
+  SliderTrackShape? get trackShape => const GappedSliderTrackShape();
+
+  @override
+  SliderComponentShape? get overlayShape => const RoundSliderOverlayShape();
+
+  @override
+  SliderTickMarkShape? get tickMarkShape => const RoundSliderTickMarkShape(tickMarkRadius: 4.0 / 2);
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider.dart:2132]: Should we put this number in a constant?

File: packages/flutter/lib/src/material/slider.dart (hunk 6 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/slider.dart:2054]: Should this number be kept in a constant?

File: packages/flutter/lib/src/material/slider.dart [hunk 9]
```diff
@@ -2054,12 +2109,48 @@ class _SliderDefaultsM3 extends SliderThemeData {
...
+
+  @override
+  MaterialStateProperty<Size?>? get barThumbSize =>
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider.dart:2135]: Multiline `=>` nit here as well.

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/material/slider_theme.dart:598]: I agree. I did not realize this theme data class was so large. There are a lot of parameters here, so having a clear name would be ideal.

[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/material/slider_theme.dart:598]: Maybe (in a separate change) we can look at this, and maybe group the parameters together based on the type of slider they apply to.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 10]
```diff
@@ -588,6 +590,20 @@ class SliderThemeData with Diagnosticable {
...
+
+  /// The size of the gap between the active and inactive tracks of the [GappedSliderTrackShape].
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider_theme.dart:600]: Is this only relevant for the M3 slider? Or both? We should check that the docs clearly state when/how a parameter/behavior/etc applies for M2 versus M3

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 11]
```diff
@@ -3556,3 +3583,408 @@ class _DropSliderValueIndicatorPathPainter {
     canvas.restore();
   }
 }
+
+/// The bar shape of a [Slider]'s thumb.
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider_theme.dart:3587]: Another thought for another change later - there's a ton of stuff in this massive file. Might be worth sending a change after this to move everything that does not relate to slider theme out of this file to a new one.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 12]
```diff
@@ -3556,3 +3583,408 @@ class _DropSliderValueIndicatorPathPainter {
...
+///
+/// If [SliderThemeData.barThumbSize] is null, then the thumb size is 4 pixels for the width
+/// and 44 pixels for the height.
+///
+/// See also:
+///
+///  * [Slider], which includes an overlay defined by this shape.
+///  * [SliderTheme], which can be used to configure the overlay shape of all
+///    sliders in a widget subtree.
+class BarSliderThumbShape extends SliderComponentShape {
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider_theme.dart:3603]: Is it default for M3?

**Review by Piinks — COMMENTED** [review]
Thank you for your patience here, I'll be much quicker on the follow up reviews. 🙏 💙

**Comment by TahaTesser** [comment]
Back from vacation, there are a lot of great feedback here and I'll address them.

File: packages/flutter/lib/src/material/slider.dart (hunk 4 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:1946]: Great idea! Removing `sliderTheme` parameter.

File: packages/flutter/lib/src/material/slider.dart (hunk 6 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:2054]: > Why the change here, do we expect it to be breaking? Is this value represented by a token?

This is updated by the  track height token which has a new value. 

```dart
class _${blockName}DefaultsM3 extends SliderThemeData {
  _${blockName}DefaultsM3(this.context)
    : super(trackHeight: ${getToken('$tokenGroup.active.track.height')});

```

File: packages/flutter/lib/src/material/slider.dart (hunk 8 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:2132]: > Should we put this number in a constant?


This value is taken from token. I think it is the correct place for this.

It actually divides the new stop indicator value by 2.
```dart
  @override
  SliderTickMarkShape? get tickMarkShape => const RoundSliderTickMarkShape(tickMarkRadius: ${getToken("$tokenGroup.stop-indicator.size")} / 2);
```

File: packages/flutter/lib/src/material/slider.dart [hunk 13]
```diff
@@ -781,20 +781,16 @@ class _SliderState extends State<Slider> with TickerProviderStateMixin {
   Widget _buildMaterialSlider(BuildContext context) {
     final ThemeData theme = Theme.of(context);
     SliderThemeData sliderTheme = SliderTheme.of(context);
-    final SliderThemeData defaults = theme.useMaterial3 ? _SliderDefaultsM3(context) : _SliderDefaultsM2(context);
-
+    final SliderThemeData defaults = theme.useMaterial3
+      ? _SliderDefaultsM3(context)
+      : _SliderDefaultsM2(context: context, sliderTheme: sliderTheme);
```
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:786]: Added new line.

File: packages/flutter/lib/src/material/slider.dart (hunk 5 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:2023]: Thank you!

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 11 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:3587]: That's right, this file contains every Slider shape and it's complex.

File: packages/flutter/lib/src/material/slider.dart (hunk 7 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:2081]: Filed https://github.com/flutter/flutter/issues/153271, will include it here

File: packages/flutter/lib/src/material/slider.dart [hunk 14]
```diff
@@ -1678,20 +1676,33 @@ class _RenderSlider extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+        trackGapSize = trackGapSize - delta / 2;
+      }
+    }
 
     _sliderTheme.trackShape!.paint(
       context,
       offset,
       parentBox: this,
-      sliderTheme: _sliderTheme,
+      sliderTheme: _sliderTheme.copyWith(trackGapSize: trackGapSize),
```
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:1699]: These are unique due to this logic where both `trackGapSize` and `thumbSize` are updated in the `_RenderSlider` widget when the thumb is interacted.

```dart
    double thumbWidth = _sliderTheme.thumbSize!.resolve(<MaterialState>{})!.width;
    final double thumbHeight = _sliderTheme.thumbSize!.resolve(<MaterialState>{})!.height;
    double trackGapSize = _sliderTheme.trackGapSize!;
    final double pressedThumbWidth = _sliderTheme.thumbSize!.resolve(<MaterialState>{ MaterialState.pressed })!.width;
    final double delta = thumbWidth - pressedThumbWidth;

    if (_active) {
      if (thumbWidth > 0.0) {
        thumbWidth = pressedThumbWidth;
      }
      if (trackGapSize > 0.0) {
        trackGapSize = trackGapSize - delta / 2;
      }
    }
```

The slider shapes do not know about these changes so i had to override these two properties so the shapes can be updated.

File: packages/flutter/lib/src/material/slider.dart [hunk 15]
```diff
@@ -1769,7 +1780,7 @@ class _RenderSlider extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
       isDiscrete: isDiscrete,
       labelPainter: _labelPainter,
       parentBox: this,
-      sliderTheme: _sliderTheme,
+      sliderTheme: _sliderTheme.copyWith(barThumbSize: MaterialStatePropertyAll<Size>(Size(thumbWidth, thumbHeight))),
```
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart]: Same explanation as  https://github.com/flutter/flutter/pull/152237#discussion_r1713621379

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 10 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:600]: All the new shapes are compatible with M2/M3.  Great suggestion

I added this documentation 
```dart
  /// The Slider defaults to [GappedSliderTrackShape] when the track shape is
  /// not specified, and the [trackGapSize] can be used to adjust the gap size.
  ///
  /// If the [ThemeData.useMaterial3] is false, then the Slider track shape
  /// defaults to [RoundedRectSliderTrackShape] and the [trackGapSize] is ignored.
  /// In this case, set the track shape to [GappedSliderTrackShape] to use the
  /// [trackGapSize].
  final double? trackGapSize;
```

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 12 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:3591]: Yes :)

**Comment by TahaTesser** [comment]
There were some conflicts which i resolved today. I'm addressing a few more API challenges such as supporting the new `thumbSize` for all shapes and also trying to figure out a better approach overriding thumb states.

**Comment by TahaTesser** [comment]
I have tried several ways to support the new `thumbSize` parameter for M2 style thumb shape. However, this turns out to be a breaking change as there are various tests depending `RoundSliderThumbShape` hard coded parameters. I also tried making them nullable but disable `RoundSliderThumbShape` cannot have nullable value in at least two tests.

https://github.com/flutter/flutter/blob/99f00a1309238338ef5a999fd86f85e200bbd044/packages/flutter/lib/src/material/slider_theme.dart#L2282-L2289

To move forward, we might need to state the new `thumbSize` theme parameter is only support the new thumb shape.

File: packages/flutter/lib/src/material/slider.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart]: I moved the pressed thumb update in a single place.
https://github.com/flutter/flutter/pull/152237/commits/2b5194b63223e8178820b9f56ae63be827d023a7

 I also tried if we could avoid this altogether but looks like there is no way to pass 'active' aka pressed state to the SliderComponentShape without breaking change.

**Comment by TahaTesser** [comment]
@QuncCccccc @Piinks 
I've addressed a bunch of feedback and commented some replies when there are breaking changes. 
Please let me know if  you agree or have any further feedback.

File: packages/flutter/lib/src/material/slider.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/material/slider.dart:1700]: Ok. Thanks for checking!

File: packages/flutter/lib/src/material/slider.dart (hunk 6 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/slider.dart:2064]: Perfect, thanks!

File: packages/flutter/lib/src/material/slider.dart (hunk 8 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/slider.dart:2143]: Thanks! Forgot about the tokens. :)

**Comment by Piinks** [comment]
> To move forward, we might need to state the new thumbSize theme parameter is only support the new M3 thumb shape.

That SGTM.

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 12 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/slider_theme.dart:3599]: Can you add that to the doc?

File: packages/flutter/lib/src/material/slider.dart [hunk 16]
```diff
@@ -782,7 +782,9 @@ class _SliderState extends State<Slider> with TickerProviderStateMixin {
   Widget _buildMaterialSlider(BuildContext context) {
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider.dart:782]: Should the docs on `Slider` be updated to reflect all of this new new? :)

Comment (Piinks) [reply packages/flutter/lib/src/material/slider.dart:782]: Bumping this, do we think there are any doc updates with this change?

**Review by Piinks — APPROVED** [review]
LGTM, kicking off Google testing 🤞 
Can you confirm the comment on `Slider` docs below?

File: packages/flutter/lib/src/material/slider.dart (hunk 16 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:782]: So sorry, i wanted to comment. Yes, I want to write this Slider doc along side the migration guide so i can mention the new stuff things in the migration guide and sync them with Slider docs.   Migration guide will have steps to restore the previous Slider components

Now that PR is approved. I will initiate migration guide draft.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 17]
```diff
@@ -3544,3 +3579,417 @@ class _DropSliderValueIndicatorPathPainter {
...
+///
+/// This is the default thumb shape for [Slider]. If [ThemeData.useMaterial3] is false,
+/// then the default thumb shape is [RoundSliderThumbShape].
+///
+/// See also:
+///
+///  * [Slider], which includes an overlay defined by this shape.
+///  * [SliderTheme], which can be used to configure the overlay shape of all
+///    sliders in a widget subtree.
+class BarSliderThumbShape extends SliderComponentShape {
```
Comment (Piinks) [inline packages/flutter/lib/src/material/slider_theme.dart:3602]: This class name conflicts with one that exists in Google testing

**Review by Piinks — COMMENTED** [review]
Let's fix the naming conflict and then I'll rerun the Google testing shard.
We might need to implement an opt in first, but it will be easier to evaluate that option once all the failures from the name conflict are gone.

I think we'll definitely want a migration guide for folks that do not want the new appearance.

I grabbed some of the screenshot test failures for you to review. Some of them looked unexpected. I can dig deeper on each in the next run, just wanted to see if any jumped out at you at first glance.

On the left is `before`, on the right is `with this change`.

Screenshot 2024-08-26 at 12 33 13 PM

Screenshot 2024-08-26 at 12 31 21 PM

Screenshot 2024-08-26 at 12 30 54 PM

Screenshot 2024-08-26 at 12 30 39 PM

Screenshot 2024-08-26 at 12 30 21 PM

Screenshot 2024-08-26 at 12 29 53 PM

**Comment by TahaTesser** [comment]
@Piinks These look to be mostly because of the updated track height of 16 pixels. I wrote a migration guide I'll file it tomorrow after Bruno looks at it. This guide includes steps to use previous track height of 4 pixels. 

The tests you shared might need to adjust their track height to get rid of these differences.

**Comment by TahaTesser** [comment]
@Piinks 
https://github.com/flutter/flutter/pull/152237/files includes steps to restore previous shapes. Could you please if the migration guide resolve the tests differences. Specifically track height.


I'm also holding on to new commits to update docs in this PR as you're checking out the tests differences in https://github.com/flutter/flutter/pull/152237#pullrequestreview-2261246024

**Comment by Piinks** [comment]
Can you push a commit that resolves the name conflict first? There are many many failures, removing those from the name conflict would help a lot. Thanks!

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 17 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:3602]: This was a we invented in the previous PR with Pierre. Naming stuff is hard :D 

Looking at this description. What If i call it `HandleThumbShape` or `VerticalBarThumbShape`?

image

**Comment by TahaTesser** [comment]
> Can you push a commit that resolves the name conflict first? There are many many failures, removing those from the name conflict would help a lot. Thanks!

Ohh thank you. I will rename this. Naming ideas https://github.com/flutter/flutter/pull/152237#discussion_r1733088462

**Comment by TahaTesser** [comment]
> Can you push a commit that resolves the name conflict first? There are many many failures, removing those from the name conflict would help a lot. Thanks!

> This class name conflicts with one that exists in Google testing


Renamed thumb shape to resolve conflict.

**Comment by TahaTesser** [comment]
Rebased to fix Flutter CI

**Comment by TahaTesser** [comment]
@Piinks Hi!  Waiting for results for https://github.com/flutter/flutter/pull/152237#pullrequestreview-2261246024 with my steps included in my migration. guide to restore the Slider to previous. If you please confirm this in fact  get rid of the diffs?  We might be able to avoid adding flag.

File: dev/tools/gen_defaults/lib/slider_template.dart [hunk 18]
```diff
@@ -36,49 +36,87 @@ class _${blockName}DefaultsM3 extends SliderThemeData {
...
+  Color? get disabledSecondaryActiveTrackColor => ${componentColor('$tokenGroup.disabled.active.track')};
 
   @override
-  Color? get activeTickMarkColor => ${componentColor('$tokenGroup.with-tick-marks.active.container')};
+  Color? get activeTickMarkColor => _colors.${getToken("$tokenGroup.stop-indicator.color-selected")};
 
   @override
-  Color? get inactiveTickMarkColor => ${componentColor('$tokenGroup.with-tick-marks.inactive.container')};
+  // TODO(tahatesser): Update this hard-coded value to use the correct token value.
+  // https://github.com/flutter/flutter/issues/153271
```
Comment (QuncCccccc) [inline dev/tools/gen_defaults/lib/slider_template.dart:46]: The fix https://github.com/flutter/flutter/pull/153722 is landed! Feel free to apply the updated tokens:)

**Comment by Piinks** [comment]
Update: Still working with affected customers in Google testing

**Comment by Piinks** [comment]
Following up on https://github.com/flutter/website/pull/11093#discussion_r1757220183, I wonder now if there are too many variants of slider that are difficult to configure with the addition of this PR.

@QuncCccccc @TahaTesser WDYT of maybe some Slider constructors specific to each kind? Especially since some properties apply to only some types of Sliders, having unique constructors could keep the number of properties exposed manageable, since only the relevant properties would be on each constructor.

**Comment by TahaTesser** [comment]
Something like` Slider.legacy` constructor? It could work, i can add private bool to check the constructor type similar to one i added for `TabBar.secondary` and configure the Slider. Do you concur @QuncCccccc ?

**Comment by Piinks** [comment]
> Something like Slider.legacy constructor?

Maybe it should correlate more with what shape/style it will result in. Inevitable, the new one here will be legacy one day. :)

**Comment by TahaTesser** [comment]
>  the new one here will be legacy one day. :)

That's true 😅 The new constructor will be for the current Slider so coming with a name to describe that, is hard.  
- They're both rounded.
- New one is has a gap but constructor for  current Slider cannot be `Slider.ungapped`, that's a bad name.
- Both have different type of thumbs but naming constructor based on thumb doesn't sound right.

**Comment by Piinks** [comment]
Good points! Maybe then the style object idea proposed in the migration guide PR is the way to go!

**Comment by QuncCccccc** [comment]
> > the new one here will be legacy one day. :)
> 
> That's true 😅 The new constructor will be for the current Slider so coming with a name to describe that, is hard.
> 
> * They're both rounded.
> * New one has a gap but constructor for  current Slider cannot be `Slider.ungapped`, that's a bad name.
> * Both have different type of thumbs but naming constructor based on thumb doesn't sound right.

How about create a `Slider.gapped()` constructor in which we have the new style? Then we don't have to handle backward compatibility problem. Since the tokens are not likely to have a big change in the future, we don't need to worry about more constructor creation:)

Adding a `sliderStyle` property sounds good to me as well. But for the existing style-related properties in Slider class, should we mark them be deprecated to avoid duplication with properties in `sliderStyle`?

**Comment by TahaTesser** [comment]
ohhh that would work if we want to move new design into a new constructor. Yes, this will keep existing Sldier untouched, we won't need migration guide at all. Genius @QuncCccccc

**Comment by QuncCccccc** [comment]
Another way that I can think of is, because given the new changes in SliderTheme, we can customize the current Slider to be a gapped Slider, then instead of providing a constructor, providing an example to show users how we can get the latest look also sounds okay:)

**Comment by TahaTesser** [comment]
> Another way that I can think of is, if given the new changes in SliderTheme, we can customize the current Slider to be a gapped Slider, then instead of providing a constructor, providing an example to show users how we can get the latest look might also work:) In this way, no backward compatibility issue,

Getting the new Slider with customization is the very first approach i took in the previous implementation of this https://github.com/flutter/flutter/pull/147783. Then we decided to go flag route, then changing defaults now.

**Comment by TahaTesser** [comment]
> Good points! Maybe then the style object idea proposed in the migration guide PR is the way to go!

I think if the future Slider updates doesn't have a gap, it might hard to keep a gapped constructor. I'm going for style object way so we can bring the customization to the widget level as well. In fact, I've an idea to simplify styling the Slider with one line to get gapped design. I'll make progress on this this week!

**Comment by TahaTesser** [comment]
This is ready for review.  This now has `Slider.year2023` flag and similar in https://github.com/flutter/flutter/pull/154817

File: packages/flutter/lib/src/material/slider.dart [hunk 19]
```diff
@@ -197,6 +197,11 @@ class Slider extends StatefulWidget {
     this.focusNode,
     this.autofocus = false,
     this.allowedInteraction,
+    @Deprecated(
+      'Use SliderTheme to customize the Slider appearance. '
+      'This feature was deprecated after v3.26.0-0.1.pre.'
+    )
+    this.year2023 = false,
```
Comment (QuncCccccc) [inline packages/flutter/lib/src/material/slider.dart:204]: Should we also make `year2023` default to true here?

**Comment by justinmc** [comment]
@TahaTesser FYI some merge conflicts here when you have a chance also.

File: packages/flutter/lib/src/material/slider.dart (hunk 19 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart:204]: Yes, we'll match this with progress indicators.

**Comment by TahaTesser** [comment]
> @TahaTesser FYI some merge conflicts here when you have a chance also.

Thanks! Looks like conflict from my other Slider PR 😁

**Comment by TahaTesser** [comment]
@QuncCccccc 
I'm working on circular progress indicator so i can publish the migration guide for both indicators. Then I'll come back to this, match the flag and resolve conflict next week here.

**Comment by justinmc** [comment]
There is still a failure but I think it's just a little indentation thing.

**Review by QuncCccccc — APPROVED** [review]
LGTM:) Thanks for the update! Will there be any migration guide to show how to keep the non-gapped Slider when `year2023` is removed from framework?

**Comment by TahaTesser** [comment]
Yes

> LGTM:) Thanks for the update! Will there be any migration guide to show how to keep the non-gapped Slider when `year2023` is removed from framework?

Yes, I'll update https://github.com/flutter/website/pull/11093 with the new flag.

**Comment by Piinks** [comment]
Looks like we are ready to merge, will follow up in migration PR. Thanks @TahaTesser for all the work on this!

**Comment by Raphael2b3** [comment]
Hello, i am on the latest flutter, and i wanted to ask when these changes are online for everyone.

flutter doktor:
[√] Flutter (Channel stable, 3.24.5, on Microsoft Windows [Version 10.0.26120.2222], locale de-DE)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.11.5)
[√] Android Studio (version 2024.2)
[√] IntelliJ IDEA Ultimate Edition (version 2024.3)
[√] Connected device (4 available)
[√] Network resources

The Slider still look like the Material 2 one. but i want to use the one from Material 3

Kind regards

**Comment by florian-obernberger** [comment]
> Hello, i am on the latest flutter, and i wanted to ask when these changes are online for everyone.
> 
> flutter doktor: [√] Flutter (Channel stable, 3.24.5, on Microsoft Windows [Version 10.0.26120.2222], locale de-DE) [√] Windows Version (Installed version of Windows is version 10 or higher) [√] Android toolchain - develop for Android devices (Android SDK version 34.0.0) [√] Chrome - develop for the web [√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.11.5) [√] Android Studio (version 2024.2) [√] IntelliJ IDEA Ultimate Edition (version 2024.3) [√] Connected device (4 available) [√] Network resources
> 
> The Slider still look like the Material 2 one. but i want to use the one from Material 3
> 
> Kind regards

In case you're still wondering: you need to set the `year2023` value to `false` in the `SliderThemeData`.

**Comment by SjdnDzikran** [comment]
Hi,

I'm trying to use the latest Material 3 `Slider` appearance. Previously, I could achieve this by setting `year2023: false`, but this now shows the following deprecation warning:

```
'year2023' is deprecated and shouldn't be used. Use SliderTheme to customize the Slider appearance. This feature was deprecated after v3.27.0-0.1.pre.
Try replacing the use of the deprecated member with the replacement.
```

What is the correct way to use the latest Material 3 `Slider` style without using the deprecated `year2023` flag? Is there an updated method or recommended approach using `SliderTheme`?

**Comment by QuncCccccc** [comment]
Hi! Setting `year2023: false` is the correct way. We marked this flag as deprecated because we wanted to encourage developers to use and gradually migrate to the latest look-and-feel Slider. When the flag is ready to be removed, we will provide dart fix and I think there will not be breaking change if you already use the latest M3 Slider:)

**Comment by pumuckelo** [comment]
@QuncCccccc but how do you use the latest M3 Slider? If I simply import the Slider component from flutter/lib/src/material it's still the old M2 design. So setting `year2023: false` in the `SliderTheme` seems to be required

**Comment by QuncCccccc** [comment]
The design of M3 has changed a lot. Originally, the M3 Slider didn't change its style much except for the default colors, so I assume that's the "old M2 design" you're referring to. In late 2023, the M3 Slider design changed to its latest version. Because of this, we updated the Slider and introduced the year2023 flag to distinguish between M3 Sliders from before late 2023 and those from after.

When we were migrating from M2 to M3, it caused many breaking changes, which was quite difficult for developers. For that reason, we decided to keep the "old" **M3** Slider design as the default. We then introduced year2023 so developers could manually enable the flag to see the latest design while avoiding those breaking changes. So yes, setting year2023 to false is required if we wanted to see the latest Slider.

**Comment by Piinks** [comment]
Hi folks, old PRs are not the best venue for feedback and discussion. If there is an issue, please file one and we can follow up, including if we need to improve the documentation.
